### DISCUSSION TITLE
How to generate ids on headings (slugs)

### DISCUSSION THREAD

**Discussion by danieldaeschle** [Milkdown/milkdown#168] [Q&A]
I tried the remark-slug plugin but it doesn't work.

Is there a way to generate it in prosemirror? I can't find any documentation for writing prosemirror plugins that updates the view.

**Comment by Saul-Mirone** [comment]
It should be a custom syntax extension, you need to not only include the remark-slug plugin, but also implement the custom prosemirror node. Something like [this chapter in document](https://milkdown.dev/#/writing-custom-plugins).

**Reply by danieldaeschle** [reply]
Your link redirects zo custom plugin. Do you mean custom plugin or syntax plugin?

That means I need a plugin that overwrites the generation of headings? Because headings are not a custom element imo.

If I understood it wrong I highly appreciate an example. Thank you for your great help!

**Reply by Saul-Mirone** [reply]
Wait, sorry I must be misunderstand this with [this extension](https://www.markdownguide.org/extended-syntax/#heading-ids).
Do you mean are that if user enter `### First Heading`, then it should generate html `<h3 id="first-heading">First Heading</h3>`.

**Reply by danieldaeschle** [reply]
Yes, exactly! Do you know how to achieve that?

**Reply by Saul-Mirone** [reply]
> Yes, exactly!

I'm working on this, I think it won't take long. You don't need to do anything except update the package version.

**Reply by danieldaeschle** [reply]
Ok, I will wait.

There seems to be a good library called github-slugger that generates the IDs.

**Reply by Saul-Mirone** [reply]
> Ok, I will wait.
> 
> There seems to be a good library called github-slugger that generates the IDs.

Just for curious, would you mind to tell that are you using this in your company's project or in your personal project?

**Reply by danieldaeschle** [reply]
It's a personal project :)
I'm trying to build a fully featured OneNote alternative and your library is awesome for this.

**Reply by Saul-Mirone** [reply]
Released in 4.14.0

**Reply by danieldaeschle** [reply]
I commented your changes and found two issues.

**Reply by Saul-Mirone** [reply]
I find it's becaused by the `toDOM` API won't be called when node updated, even worse, the content update of a node won't cause its parent node to update, which means, if I modify `# Hello` to `# Hello1`, the heading node won't be notificated.

So I think implement it directly by a prosemirror node is impossible, we need at least a prosemirror plugin like here: https://discuss.prosemirror.net/t/how-i-can-attach-attribute-with-dynamic-value-when-new-paragraph-is-inserted/751/3

So, I have to rollback changes of heading node. And you can try to implement a slug plugin if you want. If you don't have time, you can wait for me to have some band width to work on this.

**Reply by danieldaeschle** [reply]
The problem is not the time but the missing or bad documentation of prosemirror.. It's hard to understand how I should do it. Thank you for your investigation. If I get it working, I'll write it here, otherwise I appreciate your research.

**Reply by danieldaeschle** [reply]
in the source someone says:
> add an attribute with a default value of null to your paragraph

Is this possible in a plugin without providing a completely new node (e.g. overwriting the headings)?

**Reply by Saul-Mirone** [reply]
Yeah, working with prosemirror needs a lot of experience and attempts, it's pretty abstract in some way.

I think only provide a plugin isn't enough, because the dom node need to render by attribute so we also need to add an attribute for heading node.

**Reply by danieldaeschle** [reply]
Maybe milkdown needs a functionality to extend nodes. Then it might be possible.

**Reply by Saul-Mirone** [reply]
> Maybe milkdown needs a functionality to extend nodes. Then it might be possible.

That's also what I'm thinking about, I think extend nodes is a good feature.

**Reply by danieldaeschle** [reply]
Let me know if I can help you or you achieved progress :)

**Reply by danieldaeschle** [reply]
Any news?

**Reply by Saul-Mirone** [reply]
I'm doing some research about how to do it and how can we design the API, I think I'll start to coding it soon. Quite busy these days...

**Reply by Saul-Mirone** [reply]
Created  a PR for this:
https://github.com/Saul-Mirone/milkdown/pull/175

Usage:
```typescript
import { commonmark, heading } from '@milkdown/commonmark';

const myHeadingNode = heading.extend((options, utils, original) => {
 return {
    ...original,
    schema: myCustomSchema,
  };
});

editor.use(commonmark.replace(heading, myHeadingNode());
```

**Reply by danieldaeschle** [reply]
I case of a Plugin like the Heading Id which needs to be on two places. Would it be possible to solve both places with one plugin or is it only possible with applying it with two "plugins"?

**Reply by Saul-Mirone** [reply]
Are those plugins all prosemirror plugins?

**Reply by danieldaeschle** [reply]
No, I mean the link you sent above shows tw code pieces that are necessary to implement headings. It would be nice to implement this using one Milkdown plugin.

**Reply by Saul-Mirone** [reply]
Do you mean a remark plugin and a custom node plugin? For now, you have to impelment them separately. However, I'm thinking about something like **common plugin**, but I need to do some research to decide whether we need it.

**Reply by danieldaeschle** [reply]
Yep :)
Thanks
